include/micm/util/cublas_handle_singleton.hpp

```diff
@@ -0,0 +1,74 @@
+#include <memory>
+#include <mutex>
+#include <cstdlib>   // for getenv
+#include <string>    // for std::stoi
+#include <cublas_v2.h>
+#include <micm/util/cuda_util.cuh>
+
+namespace micm
+{
+  class CublasHandleSingleton {
+    public:
+      // Delete copy constructor and assignment operator to prevent copies
+      CublasHandleSingleton(const CublasHandleSingleton&) = delete;
+      CublasHandleSingleton& operator=(const CublasHandleSingleton&) = delete;
+
+      // Get the static instance of CublasHandleSingleton class
+      static CublasHandleSingleton& GetInstance()
+      {
+        const char* number_of_openmp_threads = std::getenv("OMP_NUM_THREADS");
+        const char* number_of_cuda_devices = std::getenv("NGPUS");   // This environment variable may only be set on NCAR's machine
+        if (number_of_openmp_threads != nullptr && number_of_cuda_devices != nullptr)
+        {
+            if (std::stoi(number_of_openmp_threads) > std::stoi(number_of_cuda_devices))
+            {
+            std::cout << "Mapping multiple OpenMP threads to the same GPU...\n";
+            std::lock_guard<std::mutex> lock(GetMutex()); // Lock the mutex so that only one thread will execute the following code at a time;
+                                                            // Unlock the mutex when the thread goes out of scope;
+                                                            // I think this is needed if multiple threads are mapped to the same GPU.
+            }
+        }
+        else
+        {
+            throw std::runtime_error("Please set the environment variables OMP_NUM_THREADS and CUDA_VISIBLE_DEVICES.\n");
+        }
```

Comment:
```suggestion
        auto lock = GetMutex();
```
I see no problem with getting the lock each time, even if we are in a single threaded context. Do we need all of this logic?

Reply:
Thanks @K20shores for your comment. I have added the `std::map` implementation as we discussed.